app/src/game-engine/local-engine/rules/BishopMovementRule.ts

```diff
@@ -8,38 +8,41 @@ export class BishopMovementRule extends Rule {
   public evaluate(move: Move, state: GameState): RuleEvaluationResult {
     let movingPiece = state.board.getPiece(move.source);
 
-    // If the moving piece is not a king, just delegate the evaluation to the
-    // next rule (if exists).
-    if (!movingPiece || movingPiece.type !== PieceType.Bishop) {
-      return this.nextOrInvalidResult(move, state);
-    }
+    if (!!movingPiece){
+      // If the moving piece is not a king, just delegate the evaluation to the
+      // next rule (if exists).
+      if (movingPiece.type !== PieceType.Bishop) {
+        return this.nextOrInvalidResult(move, state);
+      }
```

Comment:
There are a bunch of files with changes that seem unnecessary or at least not related with castling. Please restore all these files to simplify this review :sweat_smile:. If some general refactor is needed to implement this feature, I'd address it in a separate PR.

Reply:
seems good I will implement the flags that way.
About the unrelated changes you are right, their could be handled on a separate PR. However, the one you pointed out was a change I needed to do add because rules were not working properly, or somehow they interfered with the castling. 
The only thing I did was get sure that the pipeline logic didn't reach a rule if the piece is null. 
I can set it back, but I am afraid that may cause problems again

Reply:
The snippet:
```typescript
    if (!movingPiece || movingPiece.type !== PieceType.Bishop) {
      return this.nextOrInvalidResult(move, state);
    }
```
Ensures that the rule gets skipped if `movingPiece` is null (or am I missing something?). With this change you are introducing a short-cut in which following rules won't get executed in case of a null `movingPiece`.
If you consider we need to check for non-null moving pieces, we could add a new rule like:
```
if(!movingPiece) {
	return { valid: false };
}
```